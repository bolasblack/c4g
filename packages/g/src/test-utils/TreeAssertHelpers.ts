import {
  UnitTestTree,
  SchematicTestRunner,
} from '@angular-devkit/schematics/testing'

export function matchFileStructureSnapshot(
  creator: () => Promise<{ tree: UnitTestTree }>,
) {
  it('match file structure snapshot', async () => {
    const { tree } = await creator()
    expect(tree.files).toMatchSnapshot('file structure')
  })
}

export function matchFileContentSnapshot(
  creator: () => Promise<{ tree: UnitTestTree }>,
) {
  it('match file content snapshot', async () => {
    const { tree } = await creator()
    tree.files.forEach(file => {
      expect(tree.readContent(file)).toMatchSnapshot(`file content: ${file}`)
    })
  })
}

export function matchTasksSnapshot(
  creator: () => Promise<{ runner: SchematicTestRunner }>,
) {
  it('match tasks snapshot', async () => {
    const { runner } = await creator()
    expect(runner.tasks).toMatchSnapshot('rule tasks')
  })
}

export function matchRunResult(
  creator: () => Promise<{
    tree: UnitTestTree
    runner: SchematicTestRunner
  }>,
) {
  matchFileStructureSnapshot(creator)
  matchFileContentSnapshot(creator)
  matchTasksSnapshot(creator)
}

export function assertTreeSnapshot(
  runner: SchematicTestRunner,
  tree: UnitTestTree,
) {
  expect(tree.files).toMatchSnapshot('file structure')

  tree.files.forEach(file => {
    expect(tree.readContent(file)).toMatchSnapshot(`file content: ${file}`)
  })

  expect(runner.tasks).toMatchSnapshot('rule tasks')
}
