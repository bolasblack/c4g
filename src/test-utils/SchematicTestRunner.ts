import { Observable, of as observableOf } from 'rxjs'
import { map } from 'rxjs/operators'
import { logging } from '@angular-devkit/core'
import {
  HostTree,
  Tree,
  Collection,
  TypedSchematicContext,
  ExecutionOptions,
  Schematic,
} from '@angular-devkit/schematics'
import {
  UnitTestTree,
  SchematicTestRunner as SchematicTestRunnerBase,
} from '@angular-devkit/schematics/testing'
import { NodeModulesTestEngineHost } from '@angular-devkit/schematics/tools'

export interface CallSchematic<SchematicSchemaT> {
  (
    schematic: Schematic<{}, {}>,
    options: SchematicSchemaT,
    tree: Observable<Tree>,
    parentContext: Partial<TypedSchematicContext<{}, {}>>,
  ): Observable<Tree>
}

export class SchematicTestRunner extends SchematicTestRunnerBase {
  createContext(
    schematicName: string,
    parent?: Partial<TypedSchematicContext<{}, {}>>,
    executionOptions?: Partial<ExecutionOptions>,
  ) {
    const _collection: Collection<{}, {}> = this['_collection']
    const _logger: logging.Logger = this['_logger']
    const { engine } = this

    const schematic = _collection.createSchematic(schematicName, true)

    return engine.createContext(
      schematic,
      {
        logger: _logger,
        ...parent,
      },
      executionOptions,
    )
  }

  runSchematicAsync<SchematicSchemaT>(
    schematicName: string,
    options?: SchematicSchemaT,
    tree?: Tree,
    callSchematic?: CallSchematic<SchematicSchemaT>,
  ): Observable<UnitTestTree> {
    const _collection: Collection<{}, {}> = this['_collection']
    const _engineHost: NodeModulesTestEngineHost = this['_engineHost']
    const _logger: logging.Logger = this['_logger']

    const schematic = _collection.createSchematic(schematicName, true)
    const host = observableOf(tree || new HostTree())
    _engineHost.clearTasks()

    callSchematic = callSchematic || this._callSchematic.bind(this)

    return callSchematic!(schematic, options as any, host, {
      logger: _logger,
    }).pipe(map(tree => new UnitTestTree(tree)))
  }

  runSchematic<SchematicSchemaT extends {}>(
    schematicName: string,
    options?: SchematicSchemaT,
    tree?: Tree,
    callSchematic?: CallSchematic<SchematicSchemaT>,
  ): UnitTestTree {
    const _collection: Collection<{}, {}> = this['_collection']
    const _engineHost: NodeModulesTestEngineHost = this['_engineHost']
    const _logger: logging.Logger = this['_logger']

    const schematic = _collection.createSchematic(schematicName, true)

    let result: UnitTestTree | null = null
    let error
    const host = observableOf(tree || new HostTree())
    _engineHost.clearTasks()

    callSchematic = callSchematic || this._callSchematic.bind(this)

    callSchematic!(schematic, options as any, host, {
      logger: _logger,
    }).subscribe(t => (result = new UnitTestTree(t)), e => (error = e))

    if (error) {
      throw error
    }

    if (result === null) {
      throw new Error('Schematic is async, please use runSchematicAsync')
    }

    return result
  }

  protected _callSchematic<SchematicSchemaT extends {}>(
    schematic: Schematic<{}, {}>,
    options: SchematicSchemaT | undefined,
    tree: Observable<Tree>,
    parentContext: Partial<TypedSchematicContext<{}, {}>>,
  ) {
    return schematic.call(options || {}, tree, parentContext)
  }
}
