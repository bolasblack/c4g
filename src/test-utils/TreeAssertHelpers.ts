import { UnitTestTree } from '@angular-devkit/schematics/testing'
import { SchematicTestRunner } from '../test-utils/SchematicTestRunner'

export function assertTreeSnapshot(
  runner: SchematicTestRunner,
  tree: UnitTestTree,
) {
  // @ts-ignore
  expect(tree.files).toMatchSnapshot('file structure')
  tree.files.forEach(file => {
    // @ts-ignore
    expect(tree.readContent(file)).toMatchSnapshot(`file content: ${file}`)
  })

  // @ts-ignore
  expect(runner.tasks).toMatchSnapshot('rule tasks')
}
