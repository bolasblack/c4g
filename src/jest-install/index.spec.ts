import { Tree } from '@angular-devkit/schematics'
import { SchematicTestRunner } from '../test-utils/SchematicTestRunner'
import * as path from 'path'
import { assertTreeSnapshot } from '../test-utils/TreeAssertHelpers'
import { Options } from './index'

const collectionPath = path.join(__dirname, '../collection.json')

describe('jest-install', () => {
  it('works', () => {
    const runner = new SchematicTestRunner('schematics', collectionPath)
    const tree = runner.runSchematic<Options>('jest-install', {}, Tree.empty())
    assertTreeSnapshot(runner, tree)
  })

  it('support `react` option', () => {
    const runner = new SchematicTestRunner('schematics', collectionPath)
    const tree = runner.runSchematic<Options>(
      'jest-install',
      { react: true },
      Tree.empty(),
    )
    assertTreeSnapshot(runner, tree)
  })
})
