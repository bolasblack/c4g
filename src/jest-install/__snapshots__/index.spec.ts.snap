// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`jest-install support \`cwd\` option: file content: /hello/jest.config.js 1`] = `
"module.exports = {
  transform: {
    '^.+\\\\\\\\.js$': 'babel-jest',
    '^.+\\\\\\\\.ts$': 'ts-jest',
  },
  testMatch: [
    '**/?(*.)+(spec|test).ts',
  ],
  testPathIgnorePatterns: [
    '/node_modules/',
  ],
  moduleFileExtensions: [
    'ts',
    'js',
    'json',
    'node',
  ],
}
"
`;

exports[`jest-install support \`cwd\` option: file structure 1`] = `
Array [
  "/hello/jest.config.js",
]
`;

exports[`jest-install support \`cwd\` option: rule tasks 1`] = `
Array [
  Object {
    "name": "@c4g/node-package-install-task",
    "options": Object {
      "packageNames": Array [
        "jest",
        "jest-cli",
        "ts-jest",
        "@types/jest",
      ],
      "type": "dev",
      "workingDirectory": "",
    },
  },
]
`;

exports[`jest-install support \`react\` option: file content: /jest.config.js 1`] = `
"module.exports = {
  transform: {
    '^.+\\\\\\\\.jsx?$': 'babel-jest',
    '^.+\\\\\\\\.tsx?$': 'ts-jest',
  },
  snapshotSerializers: [
    'enzyme-to-json/serializer',
  ],
  moduleNameMapper: {
    '\\\\\\\\.(css|less|scss)$': 'identity-obj-proxy',
  },
  testMatch: [
    '**/?(*.)+(spec|test).ts?(x)',
  ],
  testPathIgnorePatterns: [
    '/node_modules/',
  ],
  moduleFileExtensions: [
    'ts',
    'tsx',
    'js',
    'jsx',
    'json',
    'node',
    'html',
  ],
}
"
`;

exports[`jest-install support \`react\` option: file structure 1`] = `
Array [
  "/jest.config.js",
]
`;

exports[`jest-install support \`react\` option: rule tasks 1`] = `
Array [
  Object {
    "name": "@c4g/node-package-install-task",
    "options": Object {
      "packageNames": Array [
        "jest",
        "jest-cli",
        "ts-jest",
        "@types/jest",
        "enzyme",
        "enzyme-to-json",
        "enzyme-adapter-react-16",
        "identity-obj-proxy",
      ],
      "type": "dev",
      "workingDirectory": "",
    },
  },
]
`;

exports[`jest-install works: file content: /jest.config.js 1`] = `
"module.exports = {
  transform: {
    '^.+\\\\\\\\.js$': 'babel-jest',
    '^.+\\\\\\\\.ts$': 'ts-jest',
  },
  testMatch: [
    '**/?(*.)+(spec|test).ts',
  ],
  testPathIgnorePatterns: [
    '/node_modules/',
  ],
  moduleFileExtensions: [
    'ts',
    'js',
    'json',
    'node',
  ],
}
"
`;

exports[`jest-install works: file structure 1`] = `
Array [
  "/jest.config.js",
]
`;

exports[`jest-install works: rule tasks 1`] = `
Array [
  Object {
    "name": "@c4g/node-package-install-task",
    "options": Object {
      "packageNames": Array [
        "jest",
        "jest-cli",
        "ts-jest",
        "@types/jest",
      ],
      "type": "dev",
      "workingDirectory": "",
    },
  },
]
`;
